File: ios/Classes/Features/QuickActions/QuickActionsFeature.swift
// =====================================================
// iOS Quick Actions Feature - Native Implementation
// =====================================================
// This file implements iOS app shortcuts (Quick Actions) functionality.
// It provides a bridge between iOS native shortcuts and Flutter Dart code.
//
// Architecture:
// 1. Flutter calls methods to create/manage shortcuts
// 2. iOS shows shortcuts when user long-presses app icon
// 3. When shortcut is tapped, AppDelegate receives the event
// 4. AppDelegate forwards via NotificationCenter to this class
// 5. This class sends shortcut data to Flutter via EventChannel
// 6. Flutter Dart code handles navigation
//
// Method Channels:
// - secure_storage_helper/quick_actions: Dart -> Native (create/clear shortcuts)
// - secure_storage_helper/quick_actions_stream: Native -> Dart (shortcut events)

import Flutter
import UIKit

public class QuickActionsFeature: NSObject, FlutterPlugin {

    // ==========================================
    // Properties & State Management
    // ==========================================

    /// Method channel for Dart to call native methods (create/clear shortcuts)
    private var channel: FlutterMethodChannel?

    /// Event channel for sending shortcut events from native to Dart
    private var eventChannel: FlutterEventChannel?

    /// Event sink to send data to Dart (nil until Dart starts listening)
    private var eventSink: FlutterEventSink?

    /// Store shortcut if event sink is not ready yet (app cold start scenario)
    private var pendingShortcutType: String?

    /// Shared instance for AppDelegate to access
    static var shared: QuickActionsFeature?

    // ==========================================
    // Plugin Registration & Setup
    // ==========================================

    /// Called by Flutter engine to register this plugin
    /// Sets up all communication channels with Dart
    public static func register(with registrar: FlutterPluginRegistrar) {
        // Create method channel for Dart -> Native calls
        let channel = FlutterMethodChannel(
            name: "secure_storage_helper/quick_actions",
            binaryMessenger: registrar.messenger())

        // Create event channel for Native -> Dart events
        let eventChannel = FlutterEventChannel(
            name: "secure_storage_helper/quick_actions_stream",
            binaryMessenger: registrar.messenger())

        let instance = QuickActionsFeature()
        instance.channel = channel
        instance.eventChannel = eventChannel

        // Register handlers
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)

        // Store reference for AppDelegate to access
        QuickActionsFeature.shared = instance

        // Listen for shortcut notifications from AppDelegate
        // This decouples AppDelegate from direct plugin access
        NotificationCenter.default.addObserver(
            instance,
            selector: #selector(instance.handleQuickActionNotification(_:)),
            name: NSNotification.Name("QuickActionPressed"),
            object: nil
        )
    }

    // ==========================================
    // Shortcut Event Handling from AppDelegate
    // ==========================================

    /// Receives shortcut notifications from AppDelegate via NotificationCenter
    /// Forwards shortcut events to Flutter Dart code
    @objc private func handleQuickActionNotification(_ notification: Notification) {
        guard let shortcutType = notification.object as? String else { return }

        print("QuickActionsFeature: Received notification for shortcut: \(shortcutType)")

        // Always store as pending first to ensure it's not lost
        pendingShortcutType = shortcutType

        // If Flutter is listening (eventSink ready), send immediately
        if let sink = eventSink {
            print("QuickActionsFeature: Sending shortcut to Flutter immediately: \(shortcutType)")

            // Add small delay to ensure Flutter UI is ready for navigation
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                sink(shortcutType)
                self.pendingShortcutType = nil  // Clear after sending
            }
        } else {
            // Flutter not ready yet (cold start), keep stored for later
            print("QuickActionsFeature: EventSink not ready, shortcut stored: \(shortcutType)")
        }
    }

    /// Static method for external access (if needed)
    public static func handleShortcut(_ shortcutType: String) {
        print("QuickActionsFeature: Static handleShortcut called with: \(shortcutType)")
        if let instance = shared {
            instance.handleShortcutAction(shortcutType)
        } else {
            print("QuickActionsFeature: No shared instance available")
        }
    }

    // ==========================================
    // Flutter Method Call Handler
    // ==========================================

    /// Handles method calls from Flutter Dart code
    /// Supported methods: setShortcutItems, clearShortcutItems
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "setShortcutItems":
            handleSetShortcutItems(call, result: result)
        case "clearShortcutItems":
            handleClearShortcutItems(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // ==========================================
    // Shortcut Management Methods
    // ==========================================

    /// Creates iOS app shortcuts from Flutter data
    /// Called when Flutter calls setShortcutItems method
    private func handleSetShortcutItems(_ call: FlutterMethodCall, result: @escaping FlutterResult)
    {
        // Validate arguments from Flutter
        guard let args = call.arguments as? [String: Any],
            let items = args["items"] as? [[String: Any]]
        else {
            result(
                FlutterError(
                    code: "INVALID_ARGUMENTS", message: "Invalid arguments for setShortcutItems",
                    details: nil))
            return
        }

        // iOS 9.0+ required for app shortcuts
        if #available(iOS 9.0, *) {
            var shortcutItems: [UIApplicationShortcutItem] = []

            // Process each shortcut item from Flutter
            for item in items {
                guard let type = item["type"] as? String,
                    let title = item["title"] as? String
                else {
                    continue
                }

                let subtitle = item["subtitle"] as? String
                let icon = item["icon"] as? String

                // Map icon names to iOS system icons
                var shortcutIcon: UIApplicationShortcutIcon?
                if let iconName = icon {
                    // Map common icon names to iOS system icons
                    switch iconName {
                    case "search":
                        shortcutIcon = UIApplicationShortcutIcon(type: .search)
                    case "compose":
                        shortcutIcon = UIApplicationShortcutIcon(type: .compose)
                    case "play":
                        shortcutIcon = UIApplicationShortcutIcon(type: .play)
                    case "pause":
                        shortcutIcon = UIApplicationShortcutIcon(type: .pause)
                    case "add":
                        shortcutIcon = UIApplicationShortcutIcon(type: .add)
                    case "location":
                        shortcutIcon = UIApplicationShortcutIcon(type: .location)
                    case "share":
                        shortcutIcon = UIApplicationShortcutIcon(type: .share)
                    case "prohibit":
                        shortcutIcon = UIApplicationShortcutIcon(type: .prohibit)
                    case "contact":
                        shortcutIcon = UIApplicationShortcutIcon(type: .contact)
                    case "home":
                        shortcutIcon = UIApplicationShortcutIcon(type: .home)
                    case "mark_location":
                        shortcutIcon = UIApplicationShortcutIcon(type: .markLocation)
                    case "favorite":
                        shortcutIcon = UIApplicationShortcutIcon(type: .favorite)
                    case "love":
                        shortcutIcon = UIApplicationShortcutIcon(type: .love)
                    case "cloud":
                        shortcutIcon = UIApplicationShortcutIcon(type: .cloud)
                    case "invitation":
                        shortcutIcon = UIApplicationShortcutIcon(type: .invitation)
                    case "confirmation":
                        shortcutIcon = UIApplicationShortcutIcon(type: .confirmation)
                    case "mail":
                        shortcutIcon = UIApplicationShortcutIcon(type: .mail)
                    case "message":
                        shortcutIcon = UIApplicationShortcutIcon(type: .message)
                    case "date":
                        shortcutIcon = UIApplicationShortcutIcon(type: .date)
                    case "time":
                        shortcutIcon = UIApplicationShortcutIcon(type: .time)
                    case "capture_photo":
                        shortcutIcon = UIApplicationShortcutIcon(type: .capturePhoto)
                    case "capture_video":
                        shortcutIcon = UIApplicationShortcutIcon(type: .captureVideo)
                    case "task":
                        shortcutIcon = UIApplicationShortcutIcon(type: .task)
                    case "task_completed":
                        shortcutIcon = UIApplicationShortcutIcon(type: .taskCompleted)
                    case "alarm":
                        shortcutIcon = UIApplicationShortcutIcon(type: .alarm)
                    case "bookmark":
                        shortcutIcon = UIApplicationShortcutIcon(type: .bookmark)
                    case "shuffle":
                        shortcutIcon = UIApplicationShortcutIcon(type: .shuffle)
                    case "audio":
                        shortcutIcon = UIApplicationShortcutIcon(type: .audio)
                    case "update":
                        shortcutIcon = UIApplicationShortcutIcon(type: .update)
                    default:
                        shortcutIcon = nil
                    }
                }

                // Create iOS shortcut item with all configured properties
                let shortcutItem = UIApplicationShortcutItem(
                    type: type,
                    localizedTitle: title,
                    localizedSubtitle: subtitle,
                    icon: shortcutIcon,
                    userInfo: nil
                )

                shortcutItems.append(shortcutItem)
            }

            // Apply shortcuts to iOS system
            UIApplication.shared.shortcutItems = shortcutItems
            result(nil)
        } else {
            result(
                FlutterError(
                    code: "UNSUPPORTED_VERSION",
                    message: "Quick actions are only supported on iOS 9.0 and above", details: nil))
        }
    }

    /// Removes all iOS app shortcuts
    /// Called when Flutter calls clearShortcutItems method
    private func handleClearShortcutItems(result: @escaping FlutterResult) {
        if #available(iOS 9.0, *) {
            UIApplication.shared.shortcutItems = []
            result(nil)
        } else {
            result(
                FlutterError(
                    code: "UNSUPPORTED_VERSION",
                    message: "Quick actions are only supported on iOS 9.0 and above", details: nil))
        }
    }
}

// ==========================================
// Event Channel Stream Handler
// ==========================================
// Handles Flutter listening for shortcut events via EventChannel

extension QuickActionsFeature: FlutterStreamHandler {

    /// Called when Flutter starts listening for shortcut events
    /// Sets up event sink and sends any pending shortcuts
    public func onListen(
        withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        print("QuickActionsFeature: EventSink connected")
        self.eventSink = events

        // Send any pending shortcut that was triggered before Flutter was ready
        if let pending = pendingShortcutType {
            print("QuickActionsFeature: Sending pending shortcut: \(pending)")
            events(pending)
            pendingShortcutType = nil
        }

        return nil
    }

    /// Called when Flutter stops listening for shortcut events
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        print("QuickActionsFeature: EventSink disconnected")
        self.eventSink = nil
        return nil
    }

    /// Internal method to handle shortcut actions
    /// Sends shortcut type to Flutter or stores if Flutter not ready
    public func handleShortcutAction(_ shortcutType: String) {
        print("QuickActionsFeature: handleShortcutAction called with: \(shortcutType)")

        if let sink = eventSink {
            // Flutter is listening, send immediately
            print("QuickActionsFeature: Sending to eventSink: \(shortcutType)")
            sink(shortcutType)
        } else {
            // Flutter not ready, store for later
            print(
                "QuickActionsFeature: No eventSink available, storing as pending: \(shortcutType)")
            pendingShortcutType = shortcutType
        }
    }
}

File: android/src/main/kotlin/com/example/secure_storage_helper/features/quick_actions/QuickActionsFeature.kt
// =====================================================
// Android Quick Actions Feature - Native Implementation
// =====================================================
// This file implements Android app shortcuts functionality.
// It provides a bridge between Android native shortcuts and Flutter Dart code.
//
// Architecture:
// 1. Flutter calls methods to create/manage shortcuts
// 2. Android shows shortcuts when user long-presses app icon
// 3. When shortcut is tapped, MainActivity receives the intent
// 4. MainActivity forwards via internal method channel to this class
// 5. This class sends shortcut data to Flutter via EventChannel
// 6. Flutter Dart code handles navigation
//
// Method Channels:
// - secure_storage_helper/quick_actions: Dart -> Native (create/clear shortcuts)
// - secure_storage_helper/quick_actions_stream: Native -> Dart (shortcut events)
// - secure_storage_helper/quick_actions_internal: MainActivity -> QuickActionsFeature

package com.example.secure_storage_helper.features.quick_actions

import android.annotation.TargetApi
import android.content.Context
import android.content.Intent
import android.content.pm.ShortcutInfo
import android.content.pm.ShortcutManager
import android.graphics.drawable.Icon
import android.os.Build
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class QuickActionsFeature : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
    
    // ==========================================
    // Properties & State Management
    // ==========================================
    
    /// Method channel for Dart to call native methods (create/clear shortcuts)
    private lateinit var channel: MethodChannel
    
    /// Event channel for sending shortcut events from native to Dart
    private lateinit var eventChannel: EventChannel
    
    /// Internal method channel for MainActivity to forward shortcuts
    private lateinit var internalChannel: MethodChannel
    
    /// Android application context
    private lateinit var context: Context
    
    /// Event sink to send data to Dart (null until Dart starts listening)
    private var eventSink: EventChannel.EventSink? = null

    companion object {
        /// Channel names - must match iOS and Dart implementations
        private const val CHANNEL_NAME = "secure_storage_helper/quick_actions"
        private const val EVENT_CHANNEL_NAME = "secure_storage_helper/quick_actions_stream"
        private const val INTERNAL_CHANNEL_NAME = "secure_storage_helper/quick_actions_internal"
        
        /// Shared instance for MainActivity access
        var instance: QuickActionsFeature? = null
    }

    // ==========================================
    // Plugin Registration & Setup
    // ==========================================
    
    /// Called by Flutter engine to register this plugin
    /// Sets up all communication channels with Dart and MainActivity
    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        // Create method channel for Dart -> Native calls
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, CHANNEL_NAME)
        
        // Create event channel for Native -> Dart events
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, EVENT_CHANNEL_NAME)
        
        // Create internal channel for MainActivity -> QuickActionsFeature communication
        internalChannel = MethodChannel(flutterPluginBinding.binaryMessenger, INTERNAL_CHANNEL_NAME)
        
        // Register handlers
        channel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(this)
        
        // Handle internal channel calls from MainActivity
        internalChannel.setMethodCallHandler { call, result ->
            when (call.method) {
                "handleShortcut" -> {
                    val shortcutType = call.arguments as? String
                    println("QuickActionsFeature: Received internal shortcut: $shortcutType")
                    if (shortcutType != null) {
                        handleShortcutAction(shortcutType)
                    }
                    result.success(null)
                }
                else -> result.notImplemented()
            }
        }
        
        context = flutterPluginBinding.applicationContext
        instance = this
    }

    /// Called when plugin is detached from Flutter engine
    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        internalChannel.setMethodCallHandler(null)
        instance = null
    }

    // ==========================================
    // Flutter Method Call Handler
    // ==========================================
    
    /// Handles method calls from Flutter Dart code
    /// Supported methods: setShortcutItems, clearShortcutItems
    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "setShortcutItems" -> {
                handleSetShortcutItems(call, result)
            }
            "clearShortcutItems" -> {
                handleClearShortcutItems(result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    // ==========================================
    // Shortcut Management Methods
    // ==========================================
    
    /// Creates Android app shortcuts from Flutter data
    /// Called when Flutter calls setShortcutItems method
    private fun handleSetShortcutItems(call: MethodCall, result: Result) {
        // Android 7.1+ (API 25) required for app shortcuts
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N_MR1) {
            result.error("UNSUPPORTED_VERSION", "App shortcuts are only supported on Android 7.1 (API 25) and above", null)
            return
        }

        val items = call.argument<List<Map<String, Any>>>("items")
        if (items == null) {
            result.error("INVALID_ARGUMENTS", "Items must be provided", null)
            return
        }

        setShortcutItems(items, result)
    }

    /// Removes all Android app shortcuts
    /// Called when Flutter calls clearShortcutItems method
    private fun handleClearShortcutItems(result: Result) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N_MR1) {
            result.error("UNSUPPORTED_VERSION", "App shortcuts are only supported on Android 7.1 (API 25) and above", null)
            return
        }

        clearShortcutItems(result)
    }

    /// Internal method to create Android shortcuts
    /// Processes Flutter shortcut data and creates native Android shortcuts
    @TargetApi(Build.VERSION_CODES.N_MR1)
    private fun setShortcutItems(items: List<Map<String, Any>>, result: Result) {
        try {
            val shortcutManager = context.getSystemService(Context.SHORTCUT_SERVICE) as ShortcutManager
            val shortcuts = mutableListOf<ShortcutInfo>()

            // Process each shortcut item from Flutter
            for (item in items) {
                val type = item["type"] as? String ?: continue
                val title = item["title"] as? String ?: continue
                val subtitle = item["subtitle"] as? String
                val iconName = item["icon"] as? String

                // Create intent that will be triggered when shortcut is tapped
                // This intent will start MainActivity with shortcut info
                val intent = Intent().apply {
                    action = Intent.ACTION_MAIN
                    setClassName(context, "com.example.secure_storage_helper_example.MainActivity")
                    putExtra("shortcut_type", type) // Pass shortcut type to MainActivity
                    flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
                }

                // Build Android shortcut with title and intent
                val builder = ShortcutInfo.Builder(context, type)
                    .setShortLabel(title)
                    .setIntent(intent)

                // Add subtitle if provided
                subtitle?.let {
                    builder.setLongLabel(it)
                }

                // Map icon name to Android system icons
                iconName?.let { name ->
                    val iconResourceId = getIconResourceId(name)
                    if (iconResourceId != 0) {
                        builder.setIcon(Icon.createWithResource(context, iconResourceId))
                    }
                }

                shortcuts.add(builder.build())
            }

            // Apply shortcuts to Android system
            shortcutManager.dynamicShortcuts = shortcuts
            result.success(null)
        } catch (e: Exception) {
            result.error("SHORTCUT_ERROR", "Failed to set shortcuts: ${e.message}", null)
        }
    }

    /// Internal method to clear all Android shortcuts
    @TargetApi(Build.VERSION_CODES.N_MR1)
    private fun clearShortcutItems(result: Result) {
        try {
            val shortcutManager = context.getSystemService(Context.SHORTCUT_SERVICE) as ShortcutManager
            shortcutManager.removeAllDynamicShortcuts()
            result.success(null)
        } catch (e: Exception) {
            result.error("SHORTCUT_ERROR", "Failed to clear shortcuts: ${e.message}", null)
        }
    }

    // ==========================================
    // Icon Mapping Helper
    // ==========================================
    
    /// Maps icon names to Android system drawable resources
    /// This ensures consistent icons across different Android versions
    private fun getIconResourceId(iconName: String): Int {
        return when (iconName) {
            "search" -> android.R.drawable.ic_menu_search
            "add" -> android.R.drawable.ic_menu_add
            "share" -> android.R.drawable.ic_menu_share
            "edit" -> android.R.drawable.ic_menu_edit
            "delete" -> android.R.drawable.ic_menu_delete
            "save" -> android.R.drawable.ic_menu_save
            "camera", "capture_photo" -> android.R.drawable.ic_menu_camera
            "gallery" -> android.R.drawable.ic_menu_gallery
            "location" -> android.R.drawable.ic_menu_mylocation
            "info" -> android.R.drawable.ic_menu_info_details
            "settings" -> android.R.drawable.ic_menu_preferences
            "help" -> android.R.drawable.ic_menu_help
            "contact" -> android.R.drawable.ic_menu_call
            "bookmark" -> android.R.drawable.ic_menu_sort_by_size
            else -> 0 // Return 0 for unknown icons (no icon will be shown)
        }
    }

    // ==========================================
    // Event Channel Stream Handler
    // ==========================================
    // Handles Flutter listening for shortcut events via EventChannel

    /// Called when Flutter starts listening for shortcut events
    /// Sets up event sink for sending shortcut events to Dart
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    /// Called when Flutter stops listening for shortcut events
    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    // ==========================================
    // Shortcut Event Handling
    // ==========================================
    
    /// Handles shortcut actions received from MainActivity
    /// Forwards shortcut events to Flutter Dart code via EventChannel
    fun handleShortcutAction(shortcutType: String) {
        println("QuickActionsFeature: Handling shortcut action: $shortcutType")
        
        eventSink?.let { sink ->
            // Flutter is listening, send shortcut event immediately
            println("QuickActionsFeature: Sending to event sink: $shortcutType")
            sink.success(shortcutType)
        } ?: run {
            // Flutter not listening yet (should not happen in normal flow)
            println("QuickActionsFeature: Event sink is null")
        }
    }
}
